### PR TITLE
Update OptionalELResolver to support method invocation

### LINKED ISSUE #27122: Expression Language 6.0: Add support for java.util.Optional. The support is not enabled by default. Users must add the new resolver OptionalELResolver to enable this support.
`https://github.com/jakartaee/expression-language/issues/176`

### PR DESCRIPTION
**Pull request by pnicolucci**
fixes #27122

### CODE REVIEW

**Comment by pnicolucci** [comment]
#libby #build

File: dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+        log("actualResult: " + actualResult);
+
+        // Without the OpionalELResolver a Java Optional is returned.
+        assertTrue("The actual result was: " + actualResult + " but was expected to be: " + expectedResult, actualResult.get().equals(expectedResult));
+    }
+
+    /**
+     * Add the OptionalELResolver to the list of ELResolver.
+     *
+     * Evaluate an Optional property that is not Empty.
```
Comment (volosied) [inline dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java]:
```suggestion
     * Add the OptionalELResolver to the list of ELResolver.
     *
     * Evaluate an Optional property that is not Empty.
     *
     *  Scenario: "If the Optional.isPresent() is true for the base object and the property is not null then the resulting value is the result of calling ELResolver.getValue(ELContext, Object, Object) with the following parameters:

    The ELContext is the current context
    The base object is the result of calling Optional.get() on the current base object
    The property object is the current property object"
    * 
```

File: dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+        log("actualResult: " + actualResult);
+
+        assertTrue("The actual result was: " + actualResult + " but was expected to be: " + expectedResult, actualResult.equals(expectedResult));
+    }
+
+    /**
+     * Add the OptionalELResolver to the list of ELResolver.
+     *
+     * Evaluate an Optional String property that is Empty.
+     *
```
Comment (volosied) [inline dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java]:
```suggestion
     * Add the OptionalELResolver to the list of ELResolver.
     *
     * Evaluate an Optional String property that is Empty.
     *
     * Scenario: "If the Optional.isEmpty() is true for the base object and the property is null then the resulting value is null."
     *
```

File: dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+
+        // With the OptionalELResolver the contents of the Java Optional are returned in this case a String.
+        assertTrue("The actual result was: " + actualResult + " but was expected to be: " + expectedResult, actualResult.equals(expectedResult));
+    }
+
+    /**
+     * Do not add the OptionalELResolver to the list of ELResolver.
+     *
+     * Evaluate an Optional String property that is Empty.
+     *
```
Comment (volosied) [inline dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java]:
```suggestion
     * Do not add the OptionalELResolver to the list of ELResolver.
     *
     * Evaluate an Optional String property that is Empty.
     *
     * Scenario: "If the Optional.isEmpty() is true for the base object and the property is not null then the resulting value is the base object (an empty Optional). " 
     *
```

File: dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java (hunk 1 above)
Comment (volosied) [reply dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java]: Asterisks are missing -- text is long and the formatting it in the suggestion is difficult.

**Review by volosied — COMMENTED** [review]
Tests look great.  I added a few small suggestions which state the scenarios in the Javadoc. 

The only thing that's puzzling me is  this scenario:

"If the Optional.isPresent() is true for the base object and the property is null then the resulting value is the result of calling Optional.get() on the base object. " 

I'm confusing myself over this. Does a test cover this?  Value is there, but property is null. If a property is not defined on bean, is it still considered null? 


edit: 

Line for reference: https://github.com/apache/tomcat/blob/eeb91237cb35dc1c9e9b5e51476df57f694bc44c/java/jakarta/el/OptionalELResolver.java#L65 

I still can't figure out what this scenario covers, but I think the current tests are adequate enough.

**Comment by volosied** [comment]
It's not necessary to add the javadoc to each test , but I though the first few were useful for easy reference (at least when I was looking over the tests).

File: dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java (hunk 1 above)
Comment (volosied) [reply dev/io.openliberty.org.apache.jasper.expressionLanguage.6.0_fat/test-applications/EL60OptionalELResolverTest.war/src/io/openliberty/el60/fat/optionalelresolver/servlets/EL60OptionalELResolverTestServlet.java]: You're right -- this scenario doesn't apply here. It applies to `testEL60OptionalELResolver_OptionalBeanProperty_WithOptionalELResolver` instead.

**Review by volosied — APPROVED** [review]
Look good to me. Thanks for the additional javadoc!

**Comment by pnicolucci** [comment]
#libby

**Comment by pnicolucci** [comment]
#build

**Comment by pnicolucci** [comment]
#build

**Comment by pnicolucci** [comment]
#build

**Comment by pnicolucci** [comment]
#libby #build

**Pull request merged**